File: client-android/app/src/main/java/stasis/client_android/api/clients/MockOAuthClient.kt
package stasis.client_android.api.clients

import org.jose4j.jwk.RsaJwkGenerator
import org.jose4j.jws.AlgorithmIdentifiers
import org.jose4j.jws.JsonWebSignature
import org.jose4j.jwt.JwtClaims
import stasis.client_android.lib.security.AccessTokenResponse
import stasis.client_android.lib.security.OAuthClient
import stasis.client_android.lib.utils.Try
import stasis.client_android.lib.utils.Try.Success

class MockOAuthClient : OAuthClient {

    override suspend fun token(scope: String?, parameters: OAuthClient.GrantParameters): Try<AccessTokenResponse> =
        Success(
            AccessTokenResponse(
                access_token = createJwt(),
                refresh_token = null,
                expires_in = 3600,
                scope = scope
            )
        )

    private val rsaKey = RsaJwkGenerator.generateJwk(2048).apply {
        keyId = "test-key"
    }

    private fun createJwt(): String {
        val claims = JwtClaims().apply {
            subject = "test-subject"
            setClaim("a", "b")
            setExpirationTimeMinutesInTheFuture(60.0f)
        }

        val jws = JsonWebSignature().apply {
            payload = claims.toJson()
            key = rsaKey.privateKey
            keyIdHeaderValue = rsaKey.keyId
            algorithmHeaderValue = AlgorithmIdentifiers.RSA_USING_SHA256
        }

        return jws.compactSerialization
    }
}

File: client-ios/app/StasisClientTests/Helpers/MockOAuthClient.swift
import Foundation
import StasisClientLib

final class MockOAuthClient: OAuthClient, @unchecked Sendable {
    private let coreOutcome: Result<AccessTokenResponse, any Error>
    private let apiOutcome: Result<AccessTokenResponse, any Error>

    init(
        coreOutcome: Result<AccessTokenResponse, any Error>,
        apiOutcome: Result<AccessTokenResponse, any Error>
    ) {
        self.coreOutcome = coreOutcome
        self.apiOutcome = apiOutcome
    }

    func token(scope: String?, parameters: GrantParameters) async -> Result<AccessTokenResponse, any Error> {
        switch parameters {
        case .clientCredentials:
            return coreOutcome
        case .resourceOwnerPasswordCredentials, .refreshToken:
            return apiOutcome
        }
    }
}

extension AccessTokenResponse {
    static func test(
        accessToken: String = "token",
        refreshToken: String? = "refresh",
        expiresIn: Int64 = 3600,
        scope: String? = nil
    ) -> AccessTokenResponse {
        AccessTokenResponse(
            accessToken: accessToken,
            refreshToken: refreshToken,
            expiresIn: expiresIn,
            scope: scope
        )
    }
}
